microraiden/microraiden/test/test_http_client.py

```diff
@@ -1,47 +1,212 @@
+            {'status_code': 200, 'headers': headers3, 'text': 'success'}
+        ])
+
+        resource = default_http_client.run('something')
+
+    # First cycle, request price.
+    request = server_mock.request_history[0]
+    assert request.path == '/something'
+    assert request.method == 'GET'
+    assert request.headers['RDN-Contract-Address'] == channel_manager_contract_address
+
+    # Second cycle, pay price based on outdated balance.
+    request = server_mock.request_history[1]
+    assert request.path == '/something'
+    assert request.method == 'GET'
+    assert request.headers['RDN-Contract-Address'] == channel_manager_contract_address
+    assert request.headers['RDN-Balance'] == '7'
+    assert request.headers['RDN-Balance-Signature']
+
+    # Third cycle, adapt new balance and pay price again.
+    request = server_mock.request_history[2]
+    assert request.path == '/something'
+    assert request.method == 'GET'
+    assert request.headers['RDN-Contract-Address'] == channel_manager_contract_address
+    assert request.headers['RDN-Balance'] == '10'
+    assert default_http_client.channel.balance == 10
+    balance_sig_hex = encode_hex(default_http_client.channel.balance_sig)
+    assert request.headers['RDN-Balance-Signature'] == balance_sig_hex
+    assert default_http_client.channel.balance_sig
+    assert resource == b'success'
+
+
+def test_full_cycle_error_500(
+        default_http_client: DefaultHTTPClient,
+        api_endpoint_address: str,
+        token_contract_address,
+        channel_manager_contract_address,
+        receiver_address
+):
+    default_http_client.initial_deposit = lambda x: x
+
+    with requests_mock.mock() as server_mock:
+        headers1 = Munch()
+        headers1.token_address = token_contract_address
+        headers1.contract_address = channel_manager_contract_address
+        headers1.receiver_address = receiver_address
+        headers1.price = '3'
+
+        headers2 = Munch()
+        headers2.cost = '3'
+
+        headers1 = HTTPHeaders.serialize(headers1)
+        headers2 = HTTPHeaders.serialize(headers2)
+
+        url = 'http://{}/something'.format(api_endpoint_address)
+        server_mock.get(url, [
+            {'status_code': 402, 'headers': headers1},
+            {'status_code': 500, 'headers': {}},
+            {'status_code': 200, 'headers': headers2, 'text': 'success'}
+        ])
+        resource = default_http_client.run('something')
+
+    # First cycle, request price.
+    request = server_mock.request_history[0]
+    assert request.path == '/something'
+    assert request.method == 'GET'
+    assert request.headers['RDN-Contract-Address'] == channel_manager_contract_address
+
+    # Second cycle, pay price but receive error.
+    request = server_mock.request_history[1]
+    assert request.path == '/something'
+    assert request.method == 'GET'
+    assert request.headers['RDN-Contract-Address'] == channel_manager_contract_address
```

Comment:
Why not use const definitions from `header.py` here?

Reply:
My semi-logical reasoning here was that I'm testing for protocol compatibility.

If the `header.py` definition is wrong then this test would still pass if using the `Munch`-based header.